File: backend/internal/analysis/analysis.go
package analysis

import (
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

var extLang = map[string]string{
	".py": "python", ".js": "javascript", ".jsx": "javascript",
	".ts": "typescript", ".tsx": "typescript", ".java": "java",
	".go": "go", ".rs": "rust", ".rb": "ruby", ".php": "php",
	".cs": "csharp", ".cpp": "cpp", ".cc": "cpp", ".c": "c",
	".sql": "sql", ".html": "html", ".css": "css", ".json": "json",
	".yml": "yaml", ".yaml": "yaml", ".md": "markdown", ".sh": "shell",
}

type Symbol struct {
	Name      string `json:"name"`
	Kind      string `json:"kind"`
	StartLine int    `json:"start_line"`
	EndLine   int    `json:"end_line"`
	Signature string `json:"signature"`
}

type AnalyzedFile struct {
	Path     string
	Language string
	Role     string
	LOC      int
	Imports  []string
	Symbols  []Symbol
}

var (
	pyImportRe = regexp.MustCompile(`(?m)^\s*(?:from\s+([\w.]+)\s+import|import\s+([\w.]+))`)
	jsImportRe = regexp.MustCompile(`(?:import\s+(?:[^'"]+\s+from\s+)?['"]([^'"]+)['"]|require\(\s*['"]([^'"]+)['"]\s*\))`)
	pyDefRe    = regexp.MustCompile(`(?m)^\s*(?:async\s+)?def\s+(\w+)\s*\((.*?)\)\s*:`)
	pyClassRe  = regexp.MustCompile(`(?m)^\s*class\s+(\w+)\s*(?:\(|:)`)
	jsFuncRe   = regexp.MustCompile(`(?:export\s+)?(?:async\s+)?function\s+(\w+)\s*\(|(?:export\s+)?const\s+(\w+)\s*=\s*(?:async\s*)?\([^)]*\)\s*=>|(?:export\s+)?class\s+(\w+)`)
	djangoURL  = regexp.MustCompile(`path\(\s*['"]([^'"]+)['"]`)
	flaskRoute = regexp.MustCompile(`@(?:\w+\.)?route\(\s*['"]([^'"]+)['"]`)
)

func DetectLanguage(path, content string) string {
	base := strings.ToLower(filepath.Base(path))
	if base == "dockerfile" || strings.HasPrefix(base, "dockerfile") {
		return "dockerfile"
	}
	ext := strings.ToLower(filepath.Ext(path))
	if lang, ok := extLang[ext]; ok {
		return lang
	}
	if strings.HasPrefix(strings.TrimSpace(content), "#!") {
		line := strings.SplitN(content, "\n", 2)[0]
		low := strings.ToLower(line)
		if strings.Contains(low, "python") {
			return "python"
		}
		if strings.Contains(low, "node") || strings.Contains(low, "bash") {
			return "shell"
		}
	}
	return "unknown"
}

func InferRole(path, language string) string {
	lower := strings.ToLower(strings.ReplaceAll(path, "\\", "/"))
	name := filepath.Base(lower)
	switch {
	case strings.Contains(lower, "migration"), strings.Contains(lower, "models.py"),
		strings.Contains(lower, "/db/"), strings.Contains(lower, "repository"), strings.Contains(lower, "prisma"):
		return "db"
	case strings.Contains(lower, "controller"), strings.Contains(lower, "views.py"),
		strings.Contains(lower, "urls.py"), strings.Contains(lower, "routes"),
		strings.Contains(lower, "api/"), strings.Contains(lower, "handlers"):
		return "api"
	case strings.Contains(lower, "component"), strings.Contains(lower, "pages/"),
		strings.Contains(lower, "frontend"), strings.Contains(lower, "ui/"),
		strings.HasSuffix(lower, ".tsx"), strings.HasSuffix(lower, ".jsx"):
		return "ui"
	case strings.Contains(lower, "worker"), strings.Contains(lower, "celery"),
		strings.Contains(lower, "task"), strings.Contains(lower, "queue"), strings.Contains(lower, "consumer"):
		return "worker"
	case strings.Contains(lower, "config"), strings.Contains(lower, "settings"),
		strings.Contains(lower, "docker"), strings.Contains(lower, "package.json"),
		strings.Contains(lower, "requirements"):
		return "config"
	case language == "html", language == "css", language == "scss":
		return "ui"
	case name == "main.py", name == "app.py", name == "index.ts", name == "index.js",
		name == "manage.py", name == "server.js", name == "main.go":
		return "entry"
	default:
		return "logic"
	}
}

func lineAt(content string, idx int) int {
	return strings.Count(content[:idx], "\n") + 1
}

func AnalyzeFile(path, abs string) AnalyzedFile {
	b, _ := os.ReadFile(abs)
	content := string(b)
	lang := DetectLanguage(path, content)
	af := AnalyzedFile{
		Path:     path,
		Language: lang,
		Role:     InferRole(path, lang),
		LOC:      strings.Count(content, "\n"),
	}
	if content != "" {
		af.LOC++
	}
	switch lang {
	case "python":
		for _, m := range pyImportRe.FindAllStringSubmatch(content, -1) {
			imp := m[1]
			if imp == "" {
				imp = m[2]
			}
			if imp != "" {
				af.Imports = append(af.Imports, imp)
			}
		}
		for _, m := range pyClassRe.FindAllStringSubmatchIndex(content, -1) {
			name := content[m[2]:m[3]]
			line := lineAt(content, m[0])
			af.Symbols = append(af.Symbols, Symbol{Name: name, Kind: "class", StartLine: line, EndLine: line, Signature: "class " + name})
		}
		for _, m := range pyDefRe.FindAllStringSubmatchIndex(content, -1) {
			name := content[m[2]:m[3]]
			args := content[m[4]:m[5]]
			line := lineAt(content, m[0])
			kind := "function"
			af.Symbols = append(af.Symbols, Symbol{Name: name, Kind: kind, StartLine: line, EndLine: line, Signature: "def " + name + "(" + args + ")"})
		}
		for _, m := range djangoURL.FindAllStringSubmatchIndex(content, -1) {
			name := content[m[2]:m[3]]
			line := lineAt(content, m[0])
			af.Symbols = append(af.Symbols, Symbol{Name: name, Kind: "endpoint", StartLine: line, EndLine: line, Signature: "path('" + name + "')"})
		}
		for _, m := range flaskRoute.FindAllStringSubmatchIndex(content, -1) {
			name := content[m[2]:m[3]]
			line := lineAt(content, m[0])
			af.Symbols = append(af.Symbols, Symbol{Name: name, Kind: "endpoint", StartLine: line, EndLine: line, Signature: "route('" + name + "')"})
		}
	case "javascript", "typescript":
		for _, m := range jsImportRe.FindAllStringSubmatch(content, -1) {
			imp := m[1]
			if imp == "" {
				imp = m[2]
			}
			if imp != "" {
				af.Imports = append(af.Imports, imp)
			}
		}
		for _, m := range jsFuncRe.FindAllStringSubmatchIndex(content, -1) {
			var name string
			for i := 1; i <= 3; i++ {
				if m[2*i] >= 0 {
					name = content[m[2*i]:m[2*i+1]]
					break
				}
			}
			if name == "" {
				continue
			}
			line := lineAt(content, m[0])
			kind := "function"
			if m[6] >= 0 {
				kind = "class"
			}
			af.Symbols = append(af.Symbols, Symbol{Name: name, Kind: kind, StartLine: line, EndLine: line, Signature: name})
		}
	}
	return af
}

func resolveImport(importName, fromPath string, all map[string]struct{}) string {
	if importName == "" {
		return ""
	}
	if strings.HasPrefix(importName, ".") {
		base := filepath.Dir(fromPath)
		rel := importName
		for strings.HasPrefix(rel, "..") {
			base = filepath.Dir(base)
			if strings.HasPrefix(rel, "../") {
				rel = rel[3:]
			} else {
				rel = rel[2:]
			}
		}
		rel = strings.TrimPrefix(rel, "./")
		cands := []string{
			filepath.ToSlash(filepath.Join(base, rel+".py")),
			filepath.ToSlash(filepath.Join(base, rel, "__init__.py")),
			filepath.ToSlash(filepath.Join(base, rel+".ts")),
			filepath.ToSlash(filepath.Join(base, rel+".tsx")),
			filepath.ToSlash(filepath.Join(base, rel+".js")),
			filepath.ToSlash(filepath.Join(base, rel, "index.ts")),
			filepath.ToSlash(filepath.Join(base, rel, "index.js")),
		}
		for _, c := range cands {
			if _, ok := all[c]; ok {
				return c
			}
		}
		return ""
	}
	dotted := strings.ReplaceAll(importName, ".", "/")
	cands := []string{dotted + ".py", dotted + "/__init__.py", dotted + ".ts", dotted + ".js", "src/" + dotted + ".ts"}
	for _, c := range cands {
		if _, ok := all[c]; ok {
			return c
		}
	}
	for p := range all {
		if strings.HasSuffix(p, "/"+dotted+".py") || strings.HasSuffix(p, "/"+dotted+".ts") {
			return p
		}
	}
	return ""
}

func BuildGraph(analyzed []AnalyzedFile) (nodes []map[string]any, edges []map[string]any) {
	all := map[string]struct{}{}
	for _, a := range analyzed {
		all[a.Path] = struct{}{}
	}
	for _, a := range analyzed {
		nodes = append(nodes, map[string]any{
			"key": a.Path, "label": filepath.Base(a.Path), "kind": a.Role,
			"metadata": map[string]any{"language": a.Language, "role": a.Role, "loc": a.LOC},
		})
	}
	seen := map[string]struct{}{}
	for _, a := range analyzed {
		for _, imp := range a.Imports {
			target := resolveImport(imp, a.Path, all)
			if target == "" || target == a.Path {
				continue
			}
			k := a.Path + "->" + target
			if _, ok := seen[k]; ok {
				continue
			}
			seen[k] = struct{}{}
			edges = append(edges, map[string]any{
				"source": a.Path, "target": target, "edge_type": "import",
				"metadata": map[string]any{"import": imp},
			})
		}
	}
	return nodes, edges
}

func ChunkFile(a AnalyzedFile, abs string, maxChars int) []map[string]any {
	b, err := os.ReadFile(abs)
	if err != nil {
		return nil
	}
	lines := strings.Split(string(b), "\n")
	var chunks []map[string]any
	for _, sym := range a.Symbols {
		start := sym.StartLine - 1
		if start < 0 {
			start = 0
		}
		end := start + 40
		if end > len(lines) {
			end = len(lines)
		}
		body := strings.Join(lines[start:end], "\n")
		if len(body) > maxChars {
			body = body[:maxChars]
		}
		if strings.TrimSpace(body) == "" {
			continue
		}
		chunks = append(chunks, map[string]any{
			"symbol_name": sym.Name, "language": a.Language,
			"start_line": start + 1, "end_line": end, "content": body,
			"metadata": map[string]any{"kind": sym.Kind, "role": a.Role},
		})
	}
	end := 80
	if end > len(lines) {
		end = len(lines)
	}
	overview := strings.Join(lines[:end], "\n")
	if len(overview) > maxChars {
		overview = overview[:maxChars]
	}
	if strings.TrimSpace(overview) != "" {
		chunks = append(chunks, map[string]any{
			"symbol_name": filepath.Base(a.Path), "language": a.Language,
			"start_line": 1, "end_line": end, "content": overview,
			"metadata": map[string]any{"kind": "file", "role": a.Role},
		})
	}
	return chunks
}

File: backend/app/services/analysis.py
"""Simple static analysis: language, role, imports, symbols, graph, chunks."""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from pathlib import Path

EXT_LANG = {
    ".py": "python",
    ".js": "javascript",
    ".jsx": "javascript",
    ".ts": "typescript",
    ".tsx": "typescript",
    ".java": "java",
    ".go": "go",
    ".rs": "rust",
    ".rb": "ruby",
    ".php": "php",
    ".sql": "sql",
    ".html": "html",
    ".css": "css",
    ".json": "json",
    ".yml": "yaml",
    ".yaml": "yaml",
    ".md": "markdown",
    ".sh": "shell",
}

PY_IMPORT = re.compile(r"^\s*(?:from\s+([\w.]+)\s+import|import\s+([\w.]+))", re.M)
JS_IMPORT = re.compile(
    r"""(?:import\s+(?:[^'"]+\s+from\s+)?['"]([^'"]+)['"]|require\(\s*['"]([^'"]+)['"]\s*\))"""
)
PY_DEF = re.compile(r"^\s*(?:async\s+)?def\s+(\w+)\s*\((.*?)\)\s*:", re.M)
PY_CLASS = re.compile(r"^\s*class\s+(\w+)\s*(?:\(|:)", re.M)
JS_FUNC = re.compile(
    r"(?:export\s+)?(?:async\s+)?function\s+(\w+)\s*\(|"
    r"(?:export\s+)?const\s+(\w+)\s*=\s*(?:async\s*)?\([^)]*\)\s*=>|"
    r"(?:export\s+)?class\s+(\w+)"
)
DJANGO_URL = re.compile(r"""path\(\s*['"]([^'"]+)['"]""")
FLASK_ROUTE = re.compile(r"""@(?:\w+\.)?route\(\s*['"]([^'"]+)['"]""")


@dataclass
class Symbol:
    name: str
    kind: str
    start_line: int
    end_line: int
    signature: str = ""


@dataclass
class AnalyzedFile:
    path: str
    language: str
    role: str
    loc: int
    imports: list[str] = field(default_factory=list)
    symbols: list[Symbol] = field(default_factory=list)


def detect_language(path: str, content: str = "") -> str:
    name = Path(path).name.lower()
    if name == "dockerfile" or name.startswith("dockerfile"):
        return "dockerfile"
    lang = EXT_LANG.get(Path(path).suffix.lower(), "")
    if lang:
        return lang
    if content.lstrip().startswith("#!"):
        shebang = content.split("\n", 1)[0].lower()
        if "python" in shebang:
            return "python"
        if "node" in shebang or "bash" in shebang:
            return "shell"
    return "unknown"


def infer_role(path: str, language: str) -> str:
    lower = path.lower().replace("\\", "/")
    name = Path(lower).name
    if any(x in lower for x in ("migration", "models.py", "/db/", "repository", "prisma")):
        return "db"
    if any(x in lower for x in ("controller", "views.py", "urls.py", "routes", "api/", "handlers")):
        return "api"
    if any(x in lower for x in ("component", "pages/", "frontend", "ui/", ".tsx", ".jsx")):
        return "ui"
    if any(x in lower for x in ("worker", "celery", "task", "queue", "consumer")):
        return "worker"
    if any(x in lower for x in ("config", "settings", "docker", "package.json", "requirements")):
        return "config"
    if language in {"html", "css"}:
        return "ui"
    if name in {"main.py", "app.py", "index.ts", "index.js", "manage.py", "server.js", "main.go"}:
        return "entry"
    return "logic"


def _line_at(content: str, idx: int) -> int:
    return content[:idx].count("\n") + 1


def analyze_file(path: str, abs_path: Path) -> AnalyzedFile:
    try:
        content = abs_path.read_text(encoding="utf-8", errors="ignore")
    except Exception:
        content = ""
    language = detect_language(path, content)
    af = AnalyzedFile(
        path=path,
        language=language,
        role=infer_role(path, language),
        loc=content.count("\n") + (1 if content else 0),
    )
    if language == "python":
        for m in PY_IMPORT.finditer(content):
            af.imports.append(m.group(1) or m.group(2))
        for m in PY_CLASS.finditer(content):
            line = _line_at(content, m.start())
            af.symbols.append(Symbol(m.group(1), "class", line, line, f"class {m.group(1)}"))
        for m in PY_DEF.finditer(content):
            line = _line_at(content, m.start())
            af.symbols.append(Symbol(m.group(1), "function", line, line, f"def {m.group(1)}({m.group(2)})"))
        for m in DJANGO_URL.finditer(content):
            line = _line_at(content, m.start())
            af.symbols.append(Symbol(m.group(1), "endpoint", line, line, f"path('{m.group(1)}')"))
        for m in FLASK_ROUTE.finditer(content):
            line = _line_at(content, m.start())
            af.symbols.append(Symbol(m.group(1), "endpoint", line, line, f"route('{m.group(1)}')"))
    elif language in {"javascript", "typescript"}:
        for m in JS_IMPORT.finditer(content):
            af.imports.append(m.group(1) or m.group(2))
        for m in JS_FUNC.finditer(content):
            name = m.group(1) or m.group(2) or m.group(3)
            if not name:
                continue
            line = _line_at(content, m.start())
            kind = "class" if m.group(3) else "function"
            af.symbols.append(Symbol(name, kind, line, line, name))
    return af


def resolve_import(import_name: str, from_path: str, all_paths: set[str]) -> str | None:
    if not import_name:
        return None
    if import_name.startswith("."):
        base = Path(from_path).parent
        rel = import_name
        while rel.startswith(".."):
            base = base.parent
            rel = rel[3:] if rel.startswith("../") else rel[2:]
        rel = rel.lstrip("./")
        candidates = [
            str(base / f"{rel}.py"),
            str(base / rel / "__init__.py"),
            str(base / f"{rel}.ts"),
            str(base / f"{rel}.tsx"),
            str(base / f"{rel}.js"),
            str(base / rel / "index.ts"),
            str(base / rel / "index.js"),
        ]
        for c in candidates:
            norm = c.replace("\\", "/")
            if norm in all_paths:
                return norm
        return None
    dotted = import_name.replace(".", "/")
    for c in [f"{dotted}.py", f"{dotted}/__init__.py", f"{dotted}.ts", f"{dotted}.js", f"src/{dotted}.ts"]:
        if c in all_paths:
            return c
    for p in all_paths:
        if p.endswith(f"/{dotted}.py") or p.endswith(f"/{dotted}.ts"):
            return p
    return None


def build_graph(analyzed: list[AnalyzedFile]) -> tuple[list[dict], list[dict]]:
    all_paths = {a.path for a in analyzed}
    nodes = [
        {
            "key": a.path,
            "label": Path(a.path).name,
            "kind": a.role or "module",
            "metadata": {"language": a.language, "role": a.role, "loc": a.loc},
        }
        for a in analyzed
    ]
    edges: list[dict] = []
    seen: set[tuple[str, str]] = set()
    for a in analyzed:
        for imp in a.imports:
            target = resolve_import(imp, a.path, all_paths)
            if not target or target == a.path:
                continue
            key = (a.path, target)
            if key in seen:
                continue
            seen.add(key)
            edges.append({"source": a.path, "target": target, "edge_type": "import", "metadata": {"import": imp}})
    return nodes, edges


def chunk_file(analyzed: AnalyzedFile, abs_path: Path, max_chars: int = 1800) -> list[dict]:
    try:
        content = abs_path.read_text(encoding="utf-8", errors="ignore")
    except Exception:
        return []
    lines = content.splitlines()
    chunks: list[dict] = []
    for sym in analyzed.symbols:
        start = max(sym.start_line - 1, 0)
        end = min(start + 40, len(lines))
        body = "\n".join(lines[start:end])
        if len(body) > max_chars:
            body = body[:max_chars]
        if not body.strip():
            continue
        chunks.append(
            {
                "symbol_name": sym.name,
                "language": analyzed.language,
                "start_line": start + 1,
                "end_line": end,
                "content": body,
                "metadata": {"kind": sym.kind, "role": analyzed.role},
            }
        )
    overview = "\n".join(lines[:80])
    if len(overview) > max_chars:
        overview = overview[:max_chars]
    if overview.strip():
        chunks.append(
            {
                "symbol_name": Path(analyzed.path).name,
                "language": analyzed.language,
                "start_line": 1,
                "end_line": min(80, len(lines)),
                "content": overview,
                "metadata": {"kind": "file", "role": analyzed.role},
            }
        )
    return chunks
